src/main/java/com/google/sps/BarcodeImageDetector.java

```
@@ -0,0 +1,103 @@
+import com.google.zxing.common.HybridBinarizer;
+
+import javax.imageio.ImageIO;
+
+import java.awt.image.BufferedImage;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * For a barcode image, detects the code, using the Java Apache Camel 
+ * Barcode API based on the ZXing library, and returns it.
+ */
+public class BarcodeImageDetector {
+  
+  public String detect(byte[] imageBytes) throws PhotoDetectionException {
+    // Convert byte array to InputStream by wrapping the byte array into the Guava ByteSource,
+    // which then allows getting the stream.
+    InputStream imageInputStream;
+    try {
+      imageInputStream = ByteSource.wrap(imageBytes).openStream();
+    } catch (IOException exception) {
+      throw new PhotoDetectionException("Failed to convert byte array to InputStream.", exception);
+    }
+    
+    // Get the BufferedImage as the result of decoding the InputStream.
+    BufferedImage bufferedImage;
+    try {
+      bufferedImage = ImageIO.read(imageInputStream);
+    } catch (IOException e) {
+      throw new PhotoDetectionException("Failed to convert InputStream to BufferedImage.", e);
+    }
+
+    // ZXing Reader objects accept a BinaryBitmap and attempt to decode it, therefore
+    // construct the BinaryBitmap object based on the {@code bufferedImage}.
+    // The ZXing HybridBinarizer class is designed for high frequency images of barcodes 
+    // with black data on white backgrounds.
+    BinaryBitmap bitmap = new BinaryBitmap(new HybridBinarizer(
+        new BufferedImageLuminanceSource(bufferedImage)));
+
+    String barcodeText;
+    try {
+      barcodeText = decode(bitmap);
+    } catch (PhotoDetectionException e) {
+      // Try to rotate the image and try decoding again.
+      if (bitmap.getWidth() < bitmap.getHeight()) {
+        if (bitmap.isRotateSupported()) {
+          bitmap = bitmap.rotateCounterClockwise();
+        }
+      }
+      return decode(bitmap);
```

This should be handled within "decode" I think. The current implementation has the following problems:

#1 it swallows the PDE exception
#2 it has different behaviors for a rotated image.